### PR TITLE
Gram matrix support for sparse input

### PR DESCRIPTION
**Pull request by mfoerste4**
This PR adds sparse input support (CSR) for GramMatrix kernel computation. This is a requirement to enable SVM support for sparse input in [cuML issue 2197](https://github.com/rapidsai/cuml/issues/2197). 

It also adds row norm computation for CSR which is utilized for expanded L2 norm computation within RBF kernels.

Although this branch introduces a new API it is still backwards compatible with the old GramMatrix API (which is marked as deprecated).

CC @cjnolet @tfeher

### CODE REVIEW

File: cpp/test/sparse/gram.cu
Comment (tfeher) [inline cpp/test/sparse/gram.cu:168]: This is the same as for dense, right? Could we reuse?

File: cpp/include/raft/distance/detail/matrix/matrix.hpp
Comment (tfeher) [inline cpp/include/raft/distance/detail/matrix/matrix.hpp:73]: Ideally we would like to use mdspan for representing the dense matrix. Could you list the pain points that would motivate keeping a pointer based API still around?

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi Malte, thanks for the PR! I have looked at the changes of the Gram matrices. Overall it looks nice, below you will find my comments.

I have added the "braking" label to remind us that we have code in cuML that is affected by the interface changes (even though it is the detail namespace, so technically it should not be breaking).

I will share my comments about the new norm method separately.

File: cpp/test/sparse/normalize.cu
Comment (tfeher) [inline cpp/test/sparse/normalize.cu:1]: This is a copy of the old tests from norm.cu.

File: cpp/include/raft/sparse/linalg/norm.cuh
Comment (tfeher) [inline cpp/include/raft/sparse/linalg/norm.cuh:81]:
```suggestion
 * @param ia the input matrix row index array
```

File: cpp/include/raft/sparse/linalg/norm.cuh
Comment (tfeher) [inline cpp/include/raft/sparse/linalg/norm.cuh:89]: - How does the performance relate to `csr_row_normalize_l1` (and `.._max`)?
- Any other differences (apart from implementation detail)?

**Review by tfeher — CHANGES_REQUESTED** [review]
Thanks Malte for updating the PR! Here are my second batch of comments (related to csr norm).

We need to find a  way not to break cuML with this PR. Can we just keep the old implementation as on overload, and remove it once the corresponding cuML PR is accepted?

File: cpp/include/raft/distance/detail/matrix/matrix.hpp
Comment (tfeher) [reply cpp/include/raft/distance/detail/matrix/matrix.hpp:73]: - I am fine keeping this version until the sparse wrapper PR #1279 is merged, and consolidate the interface afterwards.
- Note that we have helper functions to deal with mdspan on the Python side, see usage example here
https://github.com/rapidsai/raft/blob/03e26b520ea281d9b239938eda30dd5100559e24/python/pylibraft/pylibraft/neighbors/ivf_pq/ivf_pq.pyx#L523

**Review by tfeher — APPROVED** [review]
Thanks Malte for addressing the issues! The PR looks good to me, but let's wait for feedback from @cjnolet on the interface changes.

**Comment by tfeher** [comment]
@mfoerste4 after restoring the old API, we do not contain braking changes anymore, right? Could you then remove the warning message from the PR description?

File: cpp/include/raft/distance/detail/matrix/matrix.hpp
Comment (cjnolet) [reply cpp/include/raft/distance/detail/matrix/matrix.hpp:73]: I agree with @tfeher, rather than introducing more abstractions for this, we can unify the dense and sparse APIs using overloads of the existing abstractions (with templates if needed).

We didn’t have abstractions like the mdspan available when we were building cuml so we used pointers for everything. Now that we have the proper abstractions and infrastructure in place, we should use them.

please note that the sparse API has been merged. We should be able to add overloads for both dense and sparse without having an explicit class to unify them.

**Review by cjnolet — CHANGES_REQUESTED** [review]

File: cpp/include/raft/distance/detail/matrix/matrix.hpp
Comment (cjnolet) [reply cpp/include/raft/distance/detail/matrix/matrix.hpp]: Tagging @divyegala for his thoughts here regarding public APIs and template metaprogramming techniques that might help.

Comment (tfeher) [reply cpp/include/raft/distance/detail/matrix/matrix.hpp]: > rather than introducing more abstractions for this, we can unify the dense and sparse APIs using overloads of the existing abstractions (with templates if needed).

@cjnolet Having overloads instead of on unified matrix type might make sense when we consider this PR only. When looking at how this is used from cuML, that would motivate introducing a unified sparse dense type.

Let us consider the whole call chain in https://github.com/rapidsai/cuml/pull/5273:
- Cython warppers -> SvcFit -> SmoSolver.Solve -> KernelCache.getTile -> GramMatrix.evaluate
- Cython wrappers -> SvcPredict -> KernelCache.getTile -> GramMatrix.evaluate

Having overloads on this whole call chain is not feasible. Having the matrix type as template parameter would unnecessary add extra compilation steps when all we would need to pass on some metadata about the array structures.

The simplest thing to do is to wrap the input into a unified struct as high in the call chain as possible, and unpack it only  where it is needed (only in GramMatrix). The QN solver in cuML does the same: [pack inputs into a unified sturcture at the entry point](https://github.com/rapidsai/cuml/blob/6976c8b893c57b4a8ae9a4929d06cb2dc0dc4b5e/cpp/src/glm/qn/qn.cuh#L161-L200).

If you are against a unified sparse / dense type in raft, then we could keep that in cuML, but it I think we should think about having such interface in raft.

Comment (divyegala) [reply cpp/include/raft/distance/detail/matrix/matrix.hpp]: From my understanding, the code duplication in `GramMatrix` is only at the API layer but the actual logic for each input types are separate pieces anyway. Here's my suggestion, since it's trivial to overload the API layer as that doesn't have significant impact to compile time or binary size and I don't think we need any template methods here. For `GramMatrix`:
1. Keep the dense API
2. Keep the sparse API
3. Introduce unified API. This is just a facade, we immediately dispatch to the above two

This way, we solve all issues. The cuML call chain is simplified, and for users of RAFT that may only care about either the sparse API or the dense API can call that directly without us forcing them to add another type layer of `unified` before interacting with RAFT.

Comment (cjnolet) [reply cpp/include/raft/distance/detail/matrix/matrix.hpp]: My apologies @tfeher and @mfoerste4, I meant to revisit this discussion myself and things have been pretty crazy the past couple of weeks. Please understand that my scrutiny over adding new types is only because 

1) I want to make sure that we're not exploding the number of types that a user needs work with, and 
2) making sure that we aren't exploding the number of new types each time we encounter a new pattern, unless said types can be at least moderately reusable across the codebase.

It sounds like the goal here is to have a unified base class. So far, the cython APIs in RAFT have gotten pretty tedious to code- and that's just for host/device mdspan. Adding sparse to the mid and accepting that on host and device is going to make things even more tedious. I definiely understand the reasoning here. 

Currently, the `SimpleMat` contains linear algebra implementations. If the goal here is just to have a unified base class that can accept read-only inputs, could we maybe just create a very simple class called something like `read_only_matrix_view` which we extend for `csr_read_only_matrix_view` and `dense_read_only_matrix_view`?

Comment (cjnolet) [reply cpp/include/raft/distance/detail/matrix/matrix.hpp]: > As for cuML, in addition to the read-only views we also need a data/structure-owning matrix wrapper that allows for resizing internal data structures ([ResizableCsrMatrix](https://github.com/mfoerste4/cuml/blob/078ba5dc744477e1a8c57c0799fe8af981e4c3a9/cpp/src/svm/sparse_util.cuh#L34)).

It sounds like what you need there is indeed a `structure-owning` CSR matrix. Why not use [the one](https://github.com/rapidsai/raft/blob/branch-23.04/cpp/include/raft/core/device_csr_matrix.hpp#L156) that's already in RAFT? Again, it looks like we're duplicating types, which means increased maintenance burden.

File: cpp/include/raft/core/device_mdspan.hpp
Comment (cjnolet) [inline cpp/include/raft/core/device_mdspan.hpp:276]: This feels a little misplaced. First- we should be accepting the `layout_type`, rather than a simple boolean, for row_major/col_major (and we an constrain to a contiguous/strided type inside the function itself using validation.). 

From a user's perspective, I'd most often expect that a pointer would be given the leading dimension offset added before creating an mdspan from it. I suggest we create a helper/utility function that accepts a a contiguous/strided mdspan, along w/ a leading dimension and returns a new mdspan. If this is only expected to be used for testing then maybe we could just put this utility in `gram_base` for now? If you expect this to be more broadly used, maybe we should put the utility function in `core/mdspan.hpp` so it's more broadly usable (the accessor template will carry along whether it's device/host accessible).

File: cpp/include/raft/distance/detail/kernels/gram_matrix.cuh
Comment (cjnolet) [inline cpp/include/raft/distance/detail/kernels/gram_matrix.cuh:31]: We should also be exposing these publicly so users like cuml aren't using the `detail` namespace.

File: cpp/include/raft/sparse/linalg/norm.cuh
Comment (cjnolet) [inline cpp/include/raft/sparse/linalg/norm.cuh:89]: Please accept `raft::device_resources` on public APIs and pass the `cudaStream_t` down to the `detail` namespace. Arguments should be ordered as follows:
1. `device_resources`
2. input pointers
3. output pointers
4. additional parameters

File: cpp/include/raft/distance/detail/kernels/gram_matrix.cuh
Comment (cjnolet) [inline cpp/include/raft/distance/detail/kernels/gram_matrix.cuh:316]: Do we already have utilities for this?

File: cpp/include/raft/distance/detail/kernels/gram_matrix.cuh
Comment (cjnolet) [inline cpp/include/raft/distance/detail/kernels/gram_matrix.cuh:474]: We should be exposing these operations directly through our public APIs that can accept our sparse types as arguments. Can you please move this logic to a `raft/sparse/linalg/spmm.cuh` primitive like we've done for `raft/linalg/gemm.cuh`? It could just accept pointers for now. It's on our roadmap to scrape through the sparse APIs and accept our new types everywhere.

File: cpp/include/raft/distance/detail/kernels/gram_matrix.cuh
Comment (cjnolet) [inline cpp/include/raft/distance/detail/kernels/gram_matrix.cuh:370]: Please use `raft/linalg/gemm.cuh` (preferably the mdspan-based APIs) for this instead fo invoking cublas wrappers directly.

**Review by cjnolet — CHANGES_REQUESTED** [review]
@mfoerste4 I finally got a chance to review these changes more thoroughly. Overall, I think the general design looks fine but there's a few other things here to consider a few things to expose through the public API (both for reuse and so users aren't having to use the `detail` APIs.)

File: cpp/include/raft/distance/detail/kernels/gram_matrix.cuh
Comment (cjnolet) [reply cpp/include/raft/distance/detail/kernels/gram_matrix.cuh:31]: Can you explain how this would break cuml? cuML is currently not using the detail namespace for these because we're exposing them through `raft::distance::kernels` public APIs.  Such as [here](https://github.com/rapidsai/cuml/blob/branch-23.04/cpp/src/svm/smosolver.cuh#L84)

File: cpp/include/raft/sparse/linalg/norm.cuh
Comment (cjnolet) [reply cpp/include/raft/sparse/linalg/norm.cuh:89]: All of the pointer-based (dense) public APIs are deprecated in RAFT in favor of the mdspan-based APIs. We're going to be scraping through the sparse public APIs and using the new sparse types everywhere. While I'd prefer that we no longer create pointer-based sparse APIs, I do understand that we don't yet have an alternative at the moment. Still, I'd prefer to at least keep the ordering of new APIs consistent w/ the agreed-upon ordering for consistency (the dense norm computation is both deprecated and didn't yet implement the ordering.). We should always accept `device_resources` in new APIs instead of `cudaStream_t` directly.

File: cpp/include/raft/distance/detail/kernels/gram_matrix.cuh
Comment (cjnolet) [reply cpp/include/raft/distance/detail/kernels/gram_matrix.cuh:370]: Ah, yes you are right. We still have [an issue](https://github.com/rapidsai/raft/issues/875) open for that. I'm okay using the pointer-based APIs of `raft::linalg::gemm` in the meantime. Can you reference the issue I just linked in a comment at the call site for tracking?

File: cpp/include/raft/sparse/linalg/detail/norm.cuh
Comment (cjnolet) [reply cpp/include/raft/sparse/linalg/detail/norm.cuh:204]: Given that this is almost a 1:1 w/ the `csr_row_op`, I'd prefer that we use that underneath at least. It's a pretty trivial operation but the more we reuse and centralize, the more further optimizations will propagate.

File: cpp/include/raft/core/device_mdspan.hpp
Comment (cjnolet) [reply cpp/include/raft/core/device_mdspan.hpp:276]: I guess I'm thinking of something like this:
```suggestion
auto make_device_matrix_view(
  ElementType* ptr, IndexType n_rows, IndexType n_cols, bool is_row_major, IndexType ld)
{
  IndexType stride0 = is_row_major ? (ld > 0 ? ld : n_cols) : 1;
  IndexType stride1 = is_row_major ? 1 : (ld > 0 ? ld : n_rows);

  assert(is_row_major ? stride0 >= n_cols : stride1 >= n_rows);

  matrix_extent<IndexType> extents{n_rows, n_cols};
  std::array<IndexType, 2> strides{stride0, stride1};
  using mapping_type  = typename layout_stride::template mapping<matrix_extent<IndexType>>;
  mapping_type layout = {extents, strides};

  return device_matrix_view<ElementType, IndexType, layout_stride>{ptr, layout};
}
```
```
template<typename T, typename IdxT, typename LayoutPolicy = layout_c_contiguous>
auto make_strided_device_matrix_view(IdxT nrows, IdxT ncols, IdxT stride) {
    constexpr auto is_row_major = std::is_same_v<LayoutPolicy, layout_c_contiguous>;
    IndexType stride0 = is_row_major ? (ld > 0 ? ld : n_cols) : 1;
    IndexType stride1 = is_row_major ? 1 : (ld > 0 ? ld : n_rows);   

    assert(is_row_major ? stride0 >= n_cols : stride1 >= n_rows);
    matrix_extent<IndexType> extents{n_rows, n_cols};

    auto layout = make_strided_layout(extents, std::array<IndexType, 2>{stride0, stride1});
    return device_matrix_view<ElementType, IndexType, layout_stride>{ptr, layout};
}
```

**Comment by cjnolet** [comment]
@mfoerste4 just a note- I bumped this to 23.06. I'll re-review so we can get this in quickly.

File: cpp/include/raft/distance/kernels/kernel_matrices.cuh
Comment (cjnolet) [inline cpp/include/raft/distance/kernels/kernel_matrices.cuh:37]: The reason these objects were in detail originally and only the public bits exposed via typedef (`using`) was because these kernels shouldn't be exposed to the public APIs. Also, this class-based (mostly-stateless) design for host-callable computation runs counter to all of the other APIs in RAFT. I think we've all agreed that we should flatten these APIs into a series of stateless functions that can be called and have their state passed into them, but we haven't yet prioritized that, thus I wanted to keep the code out of the docs and obscure it from being discoverable by users until we have the new "formal" API exposed. I suggest we move these back into the `detail` namespace and expose them with `using` into the public namespace but also expose the other bits we need from this API to cuml in the same way (with `using`) so that cuml isn't calling APIs in the detail namespace. 

I do apologize if my wording has been confusing in previous reviews.

File: cpp/include/raft/sparse/linalg/spmm.cuh
Comment (cjnolet) [inline cpp/include/raft/sparse/linalg/spmm.cuh:51]: Tagging @divyegala for awareness thoughts since he's starting on exposing the sparse APIs w/ our new core vocabulary.

**Review by cjnolet — CHANGES_REQUESTED** [review]
Thanks for the updates, @mfoerste4. The new `norm` and `sparse::spmm` APIs look great. I think this is almost there.

**Review by cjnolet — APPROVED** [review]
LGTM. Thanks Malte!

**Comment by cjnolet** [comment]
@mfoerste4 a few recent updates to the sparse API broke your changes here. I can make these updates and push to your banch if it's easier (since I know what changed).

**Pull request merged**
